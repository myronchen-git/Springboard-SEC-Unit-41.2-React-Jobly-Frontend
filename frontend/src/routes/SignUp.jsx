function SignUp() {
  return <section>SignUp</section>;
}

// ==================================================

export default SignUp;
