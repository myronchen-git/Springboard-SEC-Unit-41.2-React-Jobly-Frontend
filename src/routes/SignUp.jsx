import PropTypes from 'prop-types';
import { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import {
  Alert,
  Button,
  Card,
  CardBody,
  CardTitle,
  Form,
  FormGroup,
  Input,
  Label,
} from 'reactstrap';

// ==================================================

/**
 * Displays and handles the account registration form.
 *
 * @param {Object} props - React component properties.
 * @param {Function} props.signup - Does the signing up.
 */
function SignUp({ signup }) {
  const initialFormData = {
    username: '',
    password: '',
    firstName: '',
    lastName: '',
    email: '',
  };
  const [formData, setFormData] = useState(initialFormData);
  const [errorMessage, setErrorMessage] = useState(null);
  const navigate = useNavigate();

  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData((formData) => ({ ...formData, [name]: value }));
  }

  async function handleSubmit(evt) {
    evt.preventDefault();

    try {
      await signup(formData);
    } catch (err) {
      setErrorMessage(err[0]);
      return;
    }

    navigate('/');
  }

  return (
    <main className="SignUp">
      <Card>
        <CardBody>
          <CardTitle tag={'h2'}>Sign Up</CardTitle>
          <Form className="SignUp__form" onSubmit={handleSubmit}>
            <FormGroup className="text-start">
              <Label htmlFor="SignUp__input-username">
                <b>Username</b>
              </Label>
              <Input
                id="SignUp__input-username"
                type="text"
                name="username"
                value={formData.username}
                required
                onChange={handleChange}
              />
            </FormGroup>
            <FormGroup className="text-start">
              <Label htmlFor="SignUp__input-password">
                <b>Password</b>
              </Label>
              <Input
                id="SignUp__input-password"
                type="password"
                name="password"
                value={formData.password}
                required
                onChange={handleChange}
              />
            </FormGroup>
            <FormGroup className="text-start">
              <Label htmlFor="SignUp__input-firstName">
                <b>First Name</b>
              </Label>
              <Input
                id="SignUp__input-firstName"
                type="text"
                name="firstName"
                value={formData.firstName}
                required
                onChange={handleChange}
              />
            </FormGroup>
            <FormGroup className="text-start">
              <Label htmlFor="SignUp__input-lastName">
                <b>Last Name</b>
              </Label>
              <Input
                id="SignUp__input-lastName"
                type="text"
                name="lastName"
                value={formData.lastName}
                required
                onChange={handleChange}
              />
            </FormGroup>
            <FormGroup className="text-start">
              <Label htmlFor="SignUp__input-email">
                <b>Email</b>
              </Label>
              <Input
                id="SignUp__input-email"
                type="email"
                name="email"
                value={formData.email}
                required
                onChange={handleChange}
              />
            </FormGroup>
            {errorMessage && <Alert color="danger">{errorMessage}</Alert>}
            <Button color="light" type="submit">
              Submit
            </Button>
          </Form>
        </CardBody>
      </Card>
    </main>
  );
}

// ==================================================

SignUp.propTypes = {
  signup: PropTypes.func.isRequired,
};

export default SignUp;
