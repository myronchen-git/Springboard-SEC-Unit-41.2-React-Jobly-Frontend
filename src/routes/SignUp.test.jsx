import { act, fireEvent, render } from '@testing-library/react';
import { useNavigate } from 'react-router-dom';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import SignUp from './SignUp.jsx';

// ==================================================

vi.mock('react-router-dom');

// ==================================================

describe('SignUp', () => {
  const formData = Object.freeze({
    username: 'user1',
    password: '12345',
    firstName: 'First',
    lastName: 'Last',
    email: 'email@email.com',
  });

  const mockSignup = vi.fn();
  const mockNavigate = vi.fn();
  useNavigate.mockReturnValue(mockNavigate);

  beforeEach(() => {
    mockSignup.mockReset();
    mockNavigate.mockReset();
  });

  it('renders.', () => {
    render(<SignUp signup={mockSignup} />);
  });

  it('matches snapshot.', () => {
    const { asFragment } = render(<SignUp signup={mockSignup} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('can update form data.', () => {
    // Arrange
    const { getByLabelText } = render(<SignUp signup={mockSignup} />);

    const usernameInput = getByLabelText('Username');
    const passwordInput = getByLabelText('Password');
    const firstNameInput = getByLabelText('First Name');
    const lastNameInput = getByLabelText('Last Name');
    const emailInput = getByLabelText('Email');

    // Act
    fireEvent.change(usernameInput, { target: { value: formData.username } });
    fireEvent.change(passwordInput, { target: { value: formData.password } });
    fireEvent.change(firstNameInput, { target: { value: formData.firstName } });
    fireEvent.change(lastNameInput, { target: { value: formData.lastName } });
    fireEvent.change(emailInput, { target: { value: formData.email } });

    // Assert
    expect(usernameInput).toHaveValue(formData.username);
    expect(passwordInput).toHaveValue(formData.password);
    expect(firstNameInput).toHaveValue(formData.firstName);
    expect(lastNameInput).toHaveValue(formData.lastName);
    expect(emailInput).toHaveValue(formData.email);

    expect(mockSignup).not.toHaveBeenCalled();
    expect(mockNavigate).not.toHaveBeenCalled();
  });

  it('can submit form data.', async () => {
    // Arrange
    const { getByLabelText, getByText } = render(
      <SignUp signup={mockSignup} />
    );

    const usernameInput = getByLabelText('Username');
    const passwordInput = getByLabelText('Password');
    const firstNameInput = getByLabelText('First Name');
    const lastNameInput = getByLabelText('Last Name');
    const emailInput = getByLabelText('Email');
    const submitBtn = getByText('Submit');

    fireEvent.change(usernameInput, { target: { value: formData.username } });
    fireEvent.change(passwordInput, { target: { value: formData.password } });
    fireEvent.change(firstNameInput, { target: { value: formData.firstName } });
    fireEvent.change(lastNameInput, { target: { value: formData.lastName } });
    fireEvent.change(emailInput, { target: { value: formData.email } });

    // Act
    await act(async () => fireEvent.click(submitBtn));

    // Assert
    expect(mockSignup).toHaveBeenCalledOnce();
    expect(mockSignup).toHaveBeenCalledWith(formData);
    expect(mockNavigate).toHaveBeenCalledWith('/');
  });

  it('displays an error message if there is an issue signing up.', async () => {
    // Arrange
    const errorMessage = 'error';
    mockSignup.mockRejectedValue([errorMessage]);

    const { getByLabelText, getByText } = render(
      <SignUp signup={mockSignup} />
    );

    const usernameInput = getByLabelText('Username');
    const passwordInput = getByLabelText('Password');
    const firstNameInput = getByLabelText('First Name');
    const lastNameInput = getByLabelText('Last Name');
    const emailInput = getByLabelText('Email');
    const submitBtn = getByText('Submit');

    fireEvent.change(usernameInput, { target: { value: formData.username } });
    fireEvent.change(passwordInput, { target: { value: formData.password } });
    fireEvent.change(firstNameInput, { target: { value: formData.firstName } });
    fireEvent.change(lastNameInput, { target: { value: formData.lastName } });
    fireEvent.change(emailInput, { target: { value: formData.email } });

    // Act
    await act(async () => fireEvent.click(submitBtn));

    // Assert
    expect(getByText(errorMessage)).toBeVisible();

    expect(mockSignup).toHaveBeenCalledOnce();
    expect(mockSignup).toHaveBeenCalledWith(formData);
    expect(mockNavigate).not.toHaveBeenCalled();
  });
});
