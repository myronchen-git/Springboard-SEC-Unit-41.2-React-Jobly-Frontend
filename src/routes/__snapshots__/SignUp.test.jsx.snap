// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignUp > matches snapshot. 1`] = `
<DocumentFragment>
  <main
    class="SignUp"
  >
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <h2
          class="card-title"
        >
          Sign Up
        </h2>
        bound HTMLFormElement {
          "0": <input
            class="form-control"
            id="SignUp__input-username"
            name="username"
            required=""
            type="text"
            value=""
          />,
          "1": <input
            class="form-control"
            id="SignUp__input-password"
            name="password"
            required=""
            type="password"
            value=""
          />,
          "2": <input
            class="form-control"
            id="SignUp__input-firstName"
            name="firstName"
            required=""
            type="text"
            value=""
          />,
          "3": <input
            class="form-control"
            id="SignUp__input-lastName"
            name="lastName"
            required=""
            type="text"
            value=""
          />,
          "4": <input
            class="form-control"
            id="SignUp__input-email"
            name="email"
            required=""
            type="email"
            value=""
          />,
          "5": <button
            class="btn btn-light"
            type="submit"
          >
            Submit
          </button>,
        }
      </div>
    </div>
  </main>
</DocumentFragment>
`;
